### ISSUE TITLE
False-negative: Method and Member names with underscores

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17839]
From: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case
> In very rare circumstances (for example, multipart version numbers), you may need to use underscores to separate adjacent numbers, since numbers do not have upper and lower case variants.

Google style is not exact. On one hand it says `_` can be used in test methods to combine sentences and on another hand in general CamelCase explanation, `_` is allowed for numbering suffix in all names(except package and generic names).

Valid uses of underscore :
```
1. guava33_4_5
2. guavaVersion33_4_5 
3. guavaversion33_4_5 (same as 1.)
```

Invalid uses of underscore:
```
1. guava_33_4_5_ (underscore at the end)
2. guava_33_4_5   (underscore between digit and letter)
3. guava_version33_4_5 (underscore between lowercase character sequences)
```

**Note**: Underscores are not allowed in normal method names. Name has to have the numbering format part at the end. Following are all invalid

```
guava_version
guava_Version
Set_guavaVersion3345
```

Some false-negatives left to address from #17708  

```java
/** some javadoc. */
public class NamesWithUnderscores {
 
  void gradle_9_5_1() {} // Expected violation, _ between digit and letter

  void jdk_9_0_392() {} // Expected violation, _ between digit and letter

  void testing_01231(String str) {} // Expected violation, _ between digit and letter

  int jdk_8_90; // Expected violation, _ between digit and letter

  int guava_33_4_7; // Expected violation, _ between digit and letter
}
```

```
$ java -jar checkstyle-11.0.0-all.jar -c google_checks.xml NamesWithUnderscores.java 
Starting audit...
Audit done.
```

we probably need a better approach than Regex. Regex has already grown very complex.
 
https://github.com/checkstyle/checkstyle/blob/5d52248c85b0fc37ddd786ddb5354e332f8a54b8/src/main/resources/google_checks.xml#L436-L438

https://github.com/checkstyle/checkstyle/blob/5d52248c85b0fc37ddd786ddb5354e332f8a54b8/src/main/resources/google_checks.xml#L254-L256

if special check is needed, it should first determine the type of name - regular or name with numbering and then it should validate the name against the appropriate regex format, which will differ for each type of naming

Keep in mind: unless it's a non-regular name i.e name that has number formating suffix at the end or method with `@Test` annotation,  underscores are not allowed.

**Comment by MiniPiku** [comment]
Hi, I’d like to take on this issue.

From the examples provided, it looks like the current regex-based validation in `MethodName` and `MemberName` doesn’t correctly distinguish between:
- Regular camelCase names
- Names with numeric version suffixes (e.g., guava33_4_5)
- Invalid underscore usage (e.g., gradle_9_5_1, guava_33_4_7)

The existing pattern:
`(?:_[0-9]+)*`

permits `_digits` sequences without requiring that the first underscore be preceded by a digit. This leads to false negatives such as:
- gradle_9_5_1
- jdk_8_90
- guava_33_4_7

According to the Google Style Guide, underscores should only be allowed:
- In version-style numeric suffixes at the end (where digits precede the first underscore), or
- In test method names.

A single complex regex makes it hard to enforce these structural rules cleanly.
Proposed approach
- First classify the name type (regular, numeric suffix, or test method).
- Apply a dedicated validation rule for each type.
- Reject any underscore usage that doesn’t fall into the allowed categories.

Please let me know if this direction aligns with your expectations before I begin implementing the fix.
Thanks!

**Comment by vivek-0509** [comment]
@MiniPiku  PR is already open for it in its subpart issues please take some other other issue,
